### PR TITLE
layer properties to dictionary

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR will need to go in before either #654 or #511 and will provide the necessary machinery to rapidly serialize (and deserialize?) the viewer.

It is still in progress, but right now it generates an `attrs` dictionary of basic `list` / `str` / `numeric` data for a layer, and an `arrays` dictionary of `array` or long `list` of `str` or `array`, or tuple of `array` data as these will likely be handled differently downstream.

I still need to add machinery to the viewer to assemble these layer dictionaries and any larger viewer dictionaries into a big dictionary.

Once assembled in #511 this could be passed to a function that wrote it out to zarr, or in #654 a function that shared it through dask distributed (doing the correct things for `arrays` vs `attrs`)

I also want to add to this PR a `deserializer` so that in something like #654 when a subscribers receives a new dictionary it can update its viewer model easily. For the special case #511 and reading a zarr frile from disk, the viewer dictionary could be passed to an empty viewer to populate a brand new one from scratch (i.e. create layers where needed vs update them).

I also want to add tests for the individual layers that the `attrs` + `arrays` match the function signature of the layer model `__init__` and test various round trips / updates.

Once all this is done then #654 and #511 can be continued (or superseded and begun again as appropriate)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
@kevinyamauchi @jni can I get your take on something.

As part of this PR I'm trying to make sure that round trips to and back from our json representation of a layer are really nice. One problem I'm encountering though is on both the `Points` and `Shapes` layer for the properties that can vary one for each shape. This may relate to discussions in #651 and ultimately get refactored, but for now if you take a property like `edge_color` on `Points`, then in the `Points` init you can pass either a single string, or a list of strings (one color for each point). Inside the layer though `edge_color` is forced to remain a single string, corresponding to the displayed text of the dropdown widget, and the list of strings is put under the `edge_colors`, this makes the round trip hard as the property name has changed and will require lots of special casing. What I do here is to move the single string to a new property `current_edge_color` and put our list of strings back under `edge_color`. Strictly speaking this will be a breaking change for anyone working with properties on our models, but it will preserve the main api of `add_points` and bring about a lot more symmetry. Ideally we'd approach the `Text` layer #600 in the same way.

Does this choice sound good to you @kevinyamauchi @jni? If so I will continue to refactor `Shapes` in the same way. Regardless of what we choose to do for the serialization stuff I think this change is in the right direction

**Comment by jni** [comment]
tbh it's a bit early in the morning and I don't have much time to think deeply about this, BUT, to me your proposal seems like a reasonable approach, and we are still in alpha so I think it's reasonable to have such a (rather minor) API change. I would suggest though that we hold off on merging this until we are close to releasing 0.3, as we probably will have a few more 0.2.x releases with bug and minor usability fixes. We should reserve this API break for 0.3. Does that sound reasonable?

**Comment by sofroniewn** [comment]
@jni we can also wait until #651 and the revamp of `Points`, `Text`, and maybe `Annotations` :-). @kevinyamauchi suggested similar recently.

I will press on with this draft, to keep exploring what this looks like in the mean time.

**Comment by kevinyamauchi** [comment]
@sofroniewn , I agree that it is reasonable to create a new property that tracks the value of a corresponding element in the UI. I did something similar for the "new text" field in the `Text` layer in #600 and I think it works well. We should come up with a standardized way to name this type of property for all layers. I vote for something like `current_<property_name>` or `selected_<property_name>`. 

I can incorporate these changes when I do the `Points` and `Text` refactor being discussed in #651 .

Also, +1 for waiting to merge this change until 0.3.0 since it is a breaking change on the API.

**Comment by sofroniewn** [comment]
This PR now adds a `_get_state` method to each layer that returns a dictionary with all the layer properties needed to recreate the layer perfectly. This will be useful during remote sharing of the viewer, saving and sharing viewer state, and maybe making animations. I'm keeping the method private for now as this is an experimental feature we might want to keep iterating on. I also made it a function so that in the future we can pass optional keyword arguments that might do things like wrap the data in dask futures calls or prepare arrays for zarr serialization etc. All this can come later I think. I also don't do serialization of the viewer right now, that can come up in a small follow up PR afterwards.

The thing this PR adds which is most important is the `test_serialize.py` test which actually tests the round trips for all our individual layers. In the future similar tests will be added to test round trips at the viewer level.

In order to make this round trip possible though we had do a lot of renaming `edge_color->current_edge_color` and `edge_colors->edge_color`. This will cause merge with #782 and some breaking changes that will need to be explicitly documented - but note non of those breaking changes are in the main API, all are in the layer properties and are simple renames
`X-> current_X` and `Xs->X`.

I'd like to get this in as soon as possible so we can press on with more changes. Does this sound reasonable @kevinyamauchi @jni?

**Comment by kevinyamauchi** [comment]
@sofroniewn, sorry! I somehow missed this! I will take a look first thing tomorrow.

File: napari/layers/points/points.py
Comment (kevinyamauchi) [inline napari/layers/points/points.py:62]: Thanks for fixing this! I just noticed this yesterday and hadn't gotten around to making an issue.

File: napari/layers/points/tests/test_points.py
Comment (kevinyamauchi) [inline napari/layers/points/tests/test_points.py]:
```suggestion
    # With no data selected changing face color has no effect
```

File: napari/layers/tests/test_serialize.py
Comment (kevinyamauchi) [inline napari/layers/tests/test_serialize.py:7]: Nice!

**Review by kevinyamauchi — APPROVED** [review]
Thanks, @sofroniewn! This looks good to me! Unfortunately, as you mentioned, I think this will create merge conflicts with #782, but it shouldn't be too bad.

**Comment by kevinyamauchi** [comment]
Also, happy new year!

File: napari/layers/points/tests/test_points.py
Comment (sofroniewn) [reply napari/layers/points/tests/test_points.py]: fixed

**Comment by sofroniewn** [comment]
Thanks for taking a look and the approve @kevinyamauchi. I have fixed the typos. I will let sit a little for @jni to take a look, but I'd like to merge by Monday if possible :-)

File: napari/layers/image/image.py
Comment (tlambert03) [inline napari/layers/image/image.py]: maybe add `contrast_limits_range` now that it's not private?

File: napari/layers/surface/surface.py
Comment (tlambert03) [inline napari/layers/surface/surface.py]: same as above `contrast_limits_range`?  (in both cases... since it will not affect the visual representation, it's not critical, but it also probably doesn't hurt)

File: napari/layers/tests/test_serialize.py
Comment (tlambert03) [reply napari/layers/tests/test_serialize.py:7]: yeah this is great.

**Review by tlambert03 — APPROVED** [review]
this is great, and I can definitely see it being useful.  I don't mind the API change for the points layers.  Once you merge this, I'll open a small PR to make layers hashable for easy equivalence checking with `layer == layer`?

I think my main comment would be syntactical ... I slightly prefer `state.update()` over the more verbose `state['key'] = value`, for example in labels.py, it would be:
```python
    def _get_state(self):
        state = self._get_base_state()
        state.update(
            {
                'is_pyramid': self.is_pyramid,
                'num_colors': self.num_colors,
                'seed': self.seed,
                'data': self.data,
            }
        )
        return state
```

File: napari/layers/image/image.py
Comment (sofroniewn) [reply napari/layers/image/image.py]: The rule I am following for adding here or not is if it is present as a keyword arg to the layer. If we add here then we should add to the function signature of `Image` and `Surface`. I'm not sure if we want to do that or not, I think we did at one point, but that can happen later

File: napari/layers/surface/surface.py
Comment (sofroniewn) [reply napari/layers/surface/surface.py]: see comment above about `init` of layer

**Comment by sofroniewn** [comment]
Following the request from @tlambert03 I've updated to use the `update` syntax. I will merge tomorrow morning, unless hearing otherwise.

A layers hashable PR would be a great followup PR.

I'll then start work on a viewer model to dictionary PR too - and address any of the remaining viewer-model separation concerns raised by @royerloic here - https://github.com/napari/napari/issues/448#issuecomment-570745596

File: napari/layers/labels/labels.py
Comment (jni) [inline napari/layers/labels/labels.py:260]: These dictionaries feel quite redundant and brittle. I would suggest that we have some kind of global/class variable, e.g. `_SERIALIZED_ATTRS`, and then this becomes:

```python
state.update({name: getattr(self, name) for name in self._SERIALIZED_ATTRS})
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py:418]: As above

File: napari/layers/tests/test_serialize.py
Comment (jni) [inline napari/layers/tests/test_serialize.py:26]: Love this bit. =D

**Review by jni — COMMENTED** [review]
@sofroniewn sorry for the delay in reviewing, but maybe my comments can be implemented as a followup PR.  They are quite minimal — great work here. =)

File: napari/layers/labels/labels.py
Comment (tlambert03) [reply napari/layers/labels/labels.py:260]: unfortunately, there's at least [one case with colormaps](https://github.com/sofroniewn/napari/blob/65a275a168ae55fef38e4568bbf6178d7822d724/napari/layers/image/image.py#L353) where this won't work because it's not just getting the attribute.  could just change the colormap attribute though

Comment (sofroniewn) [reply napari/layers/labels/labels.py:260]: ooh nice, I like that idea. i feel like maybe we should change the colormap attribute for that reason, I've always found the tuple a little unclear. maybe I'll make a short follow-up PR

Comment (jni) [reply napari/layers/labels/labels.py:260]: sounds good. Well, the colormap itself should be saved anyway, since the name doesn't always contain all the info. Do vispy colormaps serialize, though?

Comment (sofroniewn) [reply napari/layers/labels/labels.py:260]: I don't know if vispy colormaps serialize - we might want to move to some numpy / list of numpy like representation if we can - like in #782 - then our model files would be vispy independent which I think would be really nice

**Pull request merged**
